repose-aggregator/components/filters/ip-classification/src/main/scala/org/openrepose/filters/ipclassification/IPClassificationFilter.scala

```diff
@@ -0,0 +1,153 @@
+  private var userHeaderName: String = _
+  private var userHeaderQuality: Double = _
+
+  override def init(filterConfig: FilterConfig): Unit = {
+    config = new FilterConfigHelper(filterConfig).getFilterConfig(DEFAULT_CONFIG)
+    logger.info(s"Initializing using config $config")
+    val xsdURL: URL = getClass.getResource("/META-INF/config/schema/ip-classification.xsd")
+
+    configurationService.subscribeTo(
+      filterConfig.getFilterName,
+      config,
+      xsdURL,
+      this,
+      classOf[IpClassificationConfig]
+    )
+  }
+
+  override def destroy(): Unit = configurationService.unsubscribeFrom(config, this.asInstanceOf[UpdateListener[_]])
+
+  override def doFilter(servletRequest: ServletRequest, servletResponse: ServletResponse, filterChain: FilterChain): Unit = {
+    new FilterLogicHandlerDelegate(servletRequest, servletResponse, filterChain).doFilter(this)
+  }
+
+  override def handleRequest(httpServletRequest: HttpServletRequest, httpServletResponse: ReadableHttpServletResponse): FilterDirector = {
+    logger.debug("Handling HTTP request!")
+    val filterDirector = new FilterDirectorImpl()
+
+    getClassificationLabel(httpServletRequest.getRemoteAddr).foreach { label =>
+      filterDirector.requestHeaderManager.appendHeader(groupHeaderName, label, groupHeaderQuality)
+    }
+
+    //Always set the user header name to the current IP address
+    //TODO: apparently we append users as well?
+    filterDirector.requestHeaderManager().appendHeader(userHeaderName, httpServletRequest.getRemoteAddr, userHeaderQuality)
+
+    filterDirector.setFilterAction(FilterAction.PASS) //Don't need to process a response with this filter, so ship it
+    filterDirector
+  }
+
+  def getClassificationLabel(ipAddress: String): Option[String] = {
+    import scala.collection.JavaConversions._
+    cidrList.toList.find { case (_, cidrUtil) =>
+      cidrUtil.isInRange(ipAddress)
+    } map { case (label, _) =>
+      label
+    }
+  }
+
+  override def configurationUpdated(config: IpClassificationConfig): Unit = {
+    //Update all the CIDRs
+    //Create a new list to replace the old one
+    val items = new util.ArrayList[CIDRTuple]()
+
+    import scala.collection.JavaConversions._
+    val classifications = config.getClassifications.getClassification.toList
+    classifications.foreach { classification =>
+      val label = classification.getLabel
+      def splitCIDR(javaCIDR: String): List[String] = {
+
+        Option(javaCIDR).map { cidr =>
+          cidr.split(" ").toList
+        } getOrElse {
+          List.empty[String]
+        }
+      }
+
+      splitCIDR(classification.getIpv4Cidr).map { cidr =>
+        items.add((label, new CIDRUtils(cidr)))
+      }
+      splitCIDR(classification.getIpv6Cidr).map { cidr =>
+        items.add((label, new CIDRUtils(cidr)))
+      }
+    }
+
+    //We have loaded all the config, clear it and add all the new ones
+    cidrList.clear()
+    cidrList.addAll(items)
```

Comment:
This is true. The other way to do this would be to use a read-only list, and put it in an [AtomicReference](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/AtomicReference.html). This exists in java 7 as well. That's probably the right way to do it. Then it's an atomic swap out.